order-history-service/src/main/java/io/eventuate/examples/tram/ordersandcustomers/orderhistoryservice/domain/CustomerViewRepository.java

```diff
@@ -1,41 +1,85 @@
 
-import static org.springframework.data.mongodb.core.query.Criteria.where;
-
-public class CustomerViewRepository {
+public class CustomerViewRepository extends AbstractRepository {
 
-  private MongoTemplate mongoTemplate;
-
-  public CustomerViewRepository(MongoTemplate mongoTemplate) {
-    this.mongoTemplate = mongoTemplate;
+  public CustomerViewRepository(MongoClient mongoClient) {
+    super(mongoClient, "customers");
   }
 
   public Optional<CustomerView> findById(Long customerId) {
-    return Optional.ofNullable(mongoTemplate.findById(customerId, CustomerView.class));
+    Document customerDocument = collection()
+            .find(new BasicDBObject("_id", customerId))
+            .first();
+
+    return Optional
+            .ofNullable(customerDocument)
+            .map(document -> {
+              CustomerView customerView = new CustomerView();
+
+              customerView.setId(customerId);
+              customerView.setName(customerDocument.getString("name"));
+              customerView.setCreditLimit(getMoney(customerDocument, "creditLimit"));
+
+              Document orders = customerDocument.get("orders", Document.class);
+
+              if (orders != null) {
+                orders.forEach((key, value) -> {
+                  Long orderId = Long.parseLong(key);
+
+                  Document orderDocument = (Document) value;
+
+                  OrderInfo orderInfo = new OrderInfo(orderId, getMoney(orderDocument, "orderTotal"));
+
+                  Optional
+                          .ofNullable(orderDocument.getString("state"))
+                          .map(OrderState::valueOf)
+                          .ifPresent(orderInfo::setState);
+
+                  customerView.getOrders().put(orderId, orderInfo);
+                });
+              }
+
+              return customerView;
+            });
   }
 
   public void addCustomer(Long customerId, String customerName, Money creditLimit) {
-    mongoTemplate.upsert(new Query(where("id").is(customerId)),
-            new Update().set("name", customerName).set("creditLimit", creditLimit), CustomerView.class);
+    repeatOnFailure(() -> {
+      BasicDBObject customer = new BasicDBObject()
+              .append("name", customerName)
+              .append("creditLimit", creditLimit.getAmount());
+
+      collection().findOneAndUpdate(new BasicDBObject("_id", customerId),
+              new BasicDBObject("$set", customer), upsertOptions());
+    });
   }
 
   public void addOrder(Long customerId, Long orderId, Money orderTotal) {
-    mongoTemplate.upsert(new Query(where("id").is(customerId)),
-            new Update().set("orders." + orderId, new OrderInfo(orderId, orderTotal)), CustomerView.class);
+    repeatOnFailure(() -> {
+      BasicDBObject orderInfo = new BasicDBObject()
+              .append("orderId", orderId)
+              .append("orderTotal", orderTotal.getAmount());
+
+      collection().findOneAndUpdate(new BasicDBObject("_id", customerId),
```

Comment:
This pattern is repeated throughout the code:

```
      collection().findOneAndUpdate(new BasicDBObject("_id", ID),
              new BasicDBObject("$set", NEW_VALUE),
              upsertOptions());
```

Make use of `AbstractRepository` and define a `findOneAndUpdate(ID, NEW_VALUE)` method that implements this pattern

Reply:
ok